File: atari/benchmark-atari.cpp
/**
* Copyright (c) 2017-present, Facebook, Inc.
* All rights reserved.

* This source code is licensed under the BSD-style license found in the
* LICENSE file in the root directory of this source tree.
*/

//File: benchmark-atari.cpp
//g++ -O3 benchmark-atari.cc atari_game.cc -std=c++11 `pkg-config --cflags --libs ale` -I../vendor `python-config --cflags --ldflags` -DUSE_PYBIND11 && ./a.out

#include <iostream>
#include <cstring>
#include <cstdio>
#include <limits>
#include <vector>
#include <unordered_map>
#include <set>
#include <iterator>
#include <unordered_set>
#include <queue>
using namespace std;
#include "timer.hh"
#include "atari_game.h"
#include "game_context.h"

int main() {
  ContextOptions co;
  co.num_games = 256;

  GameOptions go;
  go.rom_file = "pong.bin";
  go.frame_skip = 4;
  go.seed = 42;

  GameContext GC{co, go};

  int batchsize = 16;
  GroupStat stat;
  GC.AddCollectors(batchsize, 1, 0, stat);

  cout << "Start timing ..." << endl;
  int N = 30000;
  Timer tm;
  GC.Start();
  for (int i = 0; i < N; ++i) {
    auto v = GC.Wait(0);
    GC.Steps(v);
    if (i == 0) {
      tm.restart();
    }
  }
  GC.Stop();
  cout << "tm duration" << tm.duration() << endl;
  cout << "seconds per iter:" << tm.duration() / N << endl;
  cout << N / tm.duration() * batchsize * go.frame_skip << endl;
}

File: atari/benchmark-atari.py
# Copyright (c) 2017-present, Facebook, Inc.
# All rights reserved.
#
# This source code is licensed under the BSD-style license found in the
# LICENSE file in the root directory of this source tree.

#!/usr/bin/env python
# -*- coding: utf-8 -*-
# File: benchmark-atari.py
# Author: Yuxin Wu <ppwwyyxxc@gmail.com>

import random
import time
from time import sleep
import numpy as np
import sys

from six.moves import range
import atari_game as atari
from game_utils import initialize_game
from collections import Counter

Niter = 30000
frame_skip = 4

def benchmark(ngame):
    batchsize = 16
    co, opt = initialize_game(
        batchsize, num_games=ngame,
        frame_skip=frame_skip, T=4)
    GC = atari.GameContext(co, opt)
    start = time.time()
    sys.stderr.write("Start benchmark ngame={}...\n".format(ngame))

    stats = Counter()
    for i in range(ngame): stats[i] = 0

    for k in range(Niter):
        infos = GC.Wait(0)
        #for i in range(len(infos)):
        #    stats[infos[i].meta().id] += 1
        GC.Steps(infos)
    print(stats)
    return Niter / (time.time() - start) * frame_skip * batchsize

if __name__ == '__main__':
    for ngame in [64, 128, 256, 512, 1024]:
        fps = benchmark(ngame)
        print("NGame={}, FPS={}".format(ngame, fps))
